File: src/main/python/AoC2020_08.py
#! /usr/bin/env python3
#
# Advent of Code 2020 Day 8
#
from typing import NamedTuple
from aoc import my_aocd
from aoc.common import log
from aoc.vm import VirtualMachine, Instruction, Program


class Instruction_(NamedTuple):
    operation: str
    argument: int

    @classmethod
    def instruction(cls, in_: tuple[str, str]):
        return cls(in_[0], int(in_[1]))

    def check_valid_operation(self) -> None:
        if self.operation not in ("nop", "acc", "jmp"):
            raise ValueError("Invalid instruction operation")


def _parse(inputs: tuple[str]) -> list[Instruction_]:
    return [Instruction_.instruction(input_.split()) for input_ in inputs]


def _build_program(lines: list[Instruction_]) -> Program:
    log(lines)
    instructions = list[Instruction]()
    for line in lines:
        line.check_valid_operation()
        if line.operation == "nop":
            instructions.append(Instruction.NOP())
        elif line.operation == "acc":
            instructions.append(Instruction.ADD("ACC", line.argument))
        elif line.operation == "jmp":
            instructions.append(Instruction.JMP(line.argument))
    return Program(instructions, inf_loop_treshold=1)


def _try_program_run_with_replaced_operation(instructions: list[Instruction_],
                                             index: int,
                                             new_operation: str) -> int:
    orig_instruction = instructions[index]
    instructions[index] = Instruction_(new_operation,
                                       orig_instruction.argument)
    log(f"{orig_instruction.operation} -> {new_operation}")
    try:
        program = _build_program(instructions)
        VirtualMachine().run_program(program)
    except RuntimeError:
        instructions[index] = orig_instruction
        return None
    return program.registers["ACC"]


def part_1(inputs: tuple[str]) -> int:
    log(inputs)
    instructions = _parse(inputs)
    program = _build_program(instructions)
    try:
        VirtualMachine().run_program(program)
    except RuntimeError:
        pass
    return program.registers["ACC"]


def part_2(inputs: tuple[str]) -> int:
    log(inputs)
    instructions = _parse(inputs)
    for i, instruction in enumerate(instructions):
        if instruction.operation == "nop":
            result = _try_program_run_with_replaced_operation(
                          instructions, i, "jmp")
            if result is not None:
                return result
        elif instruction.operation == "jmp":
            result = _try_program_run_with_replaced_operation(
                          instructions, i, "nop")
            if result is not None:
                return result
        elif instruction.operation == "acc":
            log("acc: skip")


TEST = """\
nop +0
acc +1
jmp +4
acc +3
jmp -3
acc -99
acc +1
jmp -4
acc +6
""".splitlines()


def main() -> None:
    my_aocd.print_header(2020, 8)

    assert part_1(TEST) == 5
    assert part_2(TEST) == 8

    inputs = my_aocd.get_input(2020, 8, 608)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")


if __name__ == '__main__':
    main()

File: src/main/java/AoC2020_08.java
import static java.util.stream.Collectors.toList;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import com.github.pareronia.aoc.vm.Instruction;
import com.github.pareronia.aoc.vm.Program;
import com.github.pareronia.aoc.vm.VirtualMachine;
import com.github.pareronia.aoc.vm.VirtualMachine.InfiniteLoopException;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2020_08 extends AoCBase {
    
	private final List<Instruction_> instructions;
	
	private AoC2020_08(final List<String> inputs, final boolean debug) {
		super(debug);
		this.instructions = parse(inputs);
		log(inputs);
	}
	
	private final List<Instruction_> parse(final List<String> inputs) {
	    return inputs.stream()
	            .map(input -> input.split(" "))
	            .map(s -> new Instruction_(s[0], Integer.valueOf(s[1])))
	            .collect(toList());
	}
	
	public static final AoC2020_08 create(final List<String> input) {
		return new AoC2020_08(input, false);
	}

	public static final AoC2020_08 createDebug(final List<String> input) {
		return new AoC2020_08(input, true);
	}
	
	private Program buildProgram(final List<Instruction_> lines) {
	    final List<Instruction> instructions = new ArrayList<>();
	    for (final Instruction_ line : lines) {
	        if (line.operator.equals("nop")) {
	            instructions.add(Instruction.NOP());
	        } else if (line.operator.equals("acc")) {
	            instructions.add(Instruction.ADD("ACC", line.operand.longValue()));
	        } else  {
	            instructions.add(Instruction.JMP(line.operand));
	        }
        }
	    return new Program(instructions, 1);
	}
	
    private Integer tryProgramWithReplaceOperation(
            final List<Instruction_> instructions,
            final int index,
            final String newOperation
    ) {
        Objects.checkFromToIndex(index, index, instructions.size());
        final Instruction_ origInstruction = instructions.get(index);
        instructions.set(index, new Instruction_(newOperation, origInstruction.operand));
	    final Program program = buildProgram(this.instructions);
	    log(program.getInstructions());
        try {
            new VirtualMachine().runProgram(program);
        } catch (final InfiniteLoopException e) {
            instructions.set(index, origInstruction);
            return null;
        }
        log(program.getCycles());
	    return program.getRegisters().get("ACC").intValue();
	}
	
	@Override
	public Integer solvePart1() {
	    final Program program = buildProgram(this.instructions);
	    log(program.getInstructions());
        try {
            new VirtualMachine().runProgram(program);
        } catch (final InfiniteLoopException e) {
        }
        log(program.getCycles());
	    return program.getRegisters().get("ACC").intValue();
	}
	
	@Override
	public Integer solvePart2() {
        for (int i = 0; i < this.instructions.size(); i++) {
            final Instruction_ instruction = this.instructions.get(i);
            if (instruction.operator.equals("nop")) {
                final Integer result = tryProgramWithReplaceOperation(instructions, i, "jmp");
                if (result != null) {
                    return result;
                }
	        } else if (instruction.operator.equals("jmp")) {
                final Integer result = tryProgramWithReplaceOperation(instructions, i, "nop");
                if (result != null) {
                    return result;
                }
	        } else if (instruction.operator.equals("acc")) {
	            // nop
	        }
        }
	    return 0;
	}

	public static void main(final String[] args) throws Exception {
		assert AoC2020_08.createDebug(TEST).solvePart1() == 5;
		assert AoC2020_08.createDebug(TEST).solvePart2() == 8;
		
        final Puzzle puzzle = Aocd.puzzle(2020, 8);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2020_08.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2020_08.create(inputData)::solvePart2)
        );
	}

	private static final List<String> TEST = splitLines(
	        "nop +0\n" +
	        "acc +1\n" +
	        "jmp +4\n" +
	        "acc +3\n" +
	        "jmp -3\n" +
	        "acc -99\n" +
	        "acc +1\n" +
	        "jmp -4\n" +
	        "acc +6"
	);
	
	record Instruction_(String operator, Integer operand) {}
}